In [0]:
spark_df = spark.sql("""Select * from adult_csv""")

In [0]:
display(spark_df)

age workclass fnlwgt education education_num marital_status occupation relationship race sex capital_gain capital_loss hours_per_week native_country income 39 State-gov 77516 Bachelors 13 Never-married Adm-clerical Not-in-family White Male 2174 0 40 United-States <=50K 50 Self-emp-not-inc 83311 Bachelors 13 Married-civ-spouse Exec-managerial Husband White Male 0 0 13 United-States <=50K 38 Private 215646 HS-grad 9 Divorced Handlers-cleaners Not-in-family White Male 0 0 40 United-States <=50K 53 Private 234721 11th 7 Married-civ-spouse Handlers-cleaners Husband Black Male 0 0 40 United-States <=50K 28 Private 338409 Bachelors 13 Married-civ-spouse Prof-specialty Wife Black Female 0 0 40 Cuba <=50K 37 Private 284582 Masters 14 Married-civ-spouse Exec-managerial Wife White Female 0 0 40 United-States <=50K 49 Private 160187 9th 5 Married-spouse-absent Other-service Not-in-family Black Female 0 0 16 Jamaica <=50K 52 Self-emp-not-inc 209642 HS-grad 9 Married-civ-spouse Exec-managerial Husband White Male 0 0 45 United-States >50K 31 Private 45781 Masters 14 Never-married Prof-specialty Not-in-family White Female 14084 0 50 United-States >50K 42 Private 159449 Bachelors 13 Married-civ-spouse Exec-managerial Husband White Male 5178 0 40 United-States >50K 37 Private 280464 Some-college 10 Married-civ-spouse Exec-managerial Husband Black Male 0 0 80 United-States >50K 30 State-gov 141297 Bachelors 13 Married-civ-spouse Prof-specialty Husband Asian-Pac-Islander Male 0 0 40 India >50K 23 Private 122272 Bachelors 13 Never-married Adm-clerical Own-child White Female 0 0 30 United-States <=50K 32 Private 205019 Assoc-acdm 12 Never-married Sales Not-in-family Black Male 0 0 50 United-States <=50K 40 Private 121772 Assoc-voc 11 Married-civ-spouse Craft-repair Husband Asian-Pac-Islander Male 0 0 40 ? >50K 34 Private 245487 7th-8th 4 Married-civ-spouse Transport-moving Husband Amer-Indian-Eskimo Male 0 0 45 Mexico <=50K 25 Self-emp-not-inc 176756 HS-grad 9 Never-married Farming-fishing Own-child White Male 0 0 35 United-States <=50K 32 Private 186824 HS-grad 9 Never-married Machine-op-inspct Unmarried White Male 0 0 40 United-States <=50K 38 Private 28887 11th 7 Married-civ-spouse Sales Husband White Male 0 0 50 United-States <=50K 43 Self-emp-not-inc 292175 Masters 14 Divorced Exec-managerial Unmarried White Female 0 0 45 United-States >50K 40 Private 193524 Doctorate 16 Married-civ-spouse Prof-specialty Husband White Male 0 0 60 United-States >50K 54 Private 302146 HS-grad 9 Separated Other-service Unmarried Black Female 0 0 20 United-States <=50K 35 Federal-gov 76845 9th 5 Married-civ-spouse Farming-fishing Husband Black Male 0 0 40 United-States <=50K 43 Private 117037 11th 7 Married-civ-spouse Transport-moving Husband White Male 0 2042 40 United-States <=50K 59 Private 109015 HS-grad 9 Divorced Tech-support Unmarried White Female 0 0 40 United-States <=50K 56 Local-gov 216851 Bachelors 13 Married-civ-spouse Tech-support Husband White Male 0 0 40 United-States >50K 19 Private 168294 HS-grad 9 Never-married Craft-repair Own-child White Male 0 0 40 United-States <=50K 54 ? 180211 Some-college 10 Married-civ-spouse ? Husband Asian-Pac-Islander Male 0 0 60 South >50K 39 Private 367260 HS-grad 9 Divorced Exec-managerial Not-in-family White Male 0 0 80 United-States <=50K 49 Private 193366 HS-grad 9 Married-civ-spouse Craft-repair Husband White Male 0 0 40 United-States <=50K 23 Local-gov 190709 Assoc-acdm 12 Never-married Protective-serv Not-in-family White Male 0 0 52 United-States <=50K 20 Private 266015 Some-college 10 Never-married Sales Own-child Black Male 0 0 44 United-States <=50K 45 Private 386940 Bachelors 13 Divorced Exec-managerial Own-child White Male 0 1408 40 United-States <=50K 30 Federal-gov 59951 Some-college 10 Married-civ-spouse Adm-clerical Own-child White Male 0 0 40 United-States <=50K 22 State-gov 311512 Some-college 10 Married-civ-spouse Other-service Husband Black Male 0 0 15 United-States <=50K 48 Private 242406 11th 7

In [0]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt # for data visualization purposes
import seaborn as sns # for statistical data visualization
%matplotlib inline

In [0]:
import warnings

warnings.filterwarnings('ignore')

# Load dataset <a class="anchor" id="6"></a>

In [0]:
df = spark_df.toPandas()

# **Exploratory data analysis** <a class="anchor" id="7"></a>

In [0]:
# view dimensions of dataset

df.shape

Out[6]: (32561, 15)

We can see that there are 32561 instances and 15 attributes in the data set.

### View top 5 rows of dataset

In [0]:
# preview the dataset

df.head()

Out[7]:

age 
 workclass 
 fnlwgt 
 education 
 education_num 
 marital_status 
 occupation 
 relationship 
 race 
 sex 
 capital_gain 
 capital_loss 
 hours_per_week 
 native_country 
 income 
 
 
 
 
 0 
 39 
 State-gov 
 77516 
 Bachelors 
 13 
 Never-married 
 Adm-clerical 
 Not-in-family 
 White 
 Male 
 2174 
 0 
 40 
 United-States 
 <=50K 
 
 
 1 
 50 
 Self-emp-not-inc 
 83311 
 Bachelors 
 13 
 Married-civ-spouse 
 Exec-managerial 
 Husband 
 White 
 Male 
 0 
 0 
 13 
 United-States 
 <=50K 
 
 
 2 
 38 
 Private 
 215646 
 HS-grad 
 9 
 Divorced 
 Handlers-cleaners 
 Not-in-family 
 White 
 Male 
 0 
 0 
 40 
 United-States 
 <=50K 
 
 
 3 
 53 
 Private 
 234721 
 11th 
 7 
 Married-civ-spouse 
 Handlers-cleaners 
 Husband 
 Black 
 Male 
 0 
 0 
 40 
 United-States 
 <=50K 
 
 
 4 
 28 
 Private 
 338409 
 Bachelors 
 13 
 Married-civ-spouse 
 Prof-specialty 
 Wife 
 Black 
 Female 
 0 
 0 
 40 
 Cuba 
 <=50K

We can see that the column names are renamed. Now, the columns have meaningful names.

### View summary of dataset

In [0]:
cols = [i.strip() for i in df.columns]
df.columns = cols

In [0]:
# view summary of dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 # Column Non-Null Count Dtype 
--- ------ -------------- ----- 
 0 age 32561 non-null object
 1 workclass 32561 non-null object
 2 fnlwgt 32561 non-null object
 3 education 32561 non-null object
 4 education_num 32561 non-null object
 5 marital_status 32561 non-null object
 6 occupation 32561 non-null object
 7 relationship 32561 non-null object
 8 race 32561 non-null object
 9 sex 32561 non-null object
 10 capital_gain 32561 non-null object
 11 capital_loss 32561 non-null object
 12 hours_per_week 32561 non-null object
 13 native_country 32561 non-null object
 14 income 32561 non-null object
dtypes: object(15)
memory usage: 3.7+ MB

In [0]:
#Proper type conversion
df['age'] = df['age'].astype(int)
df['fnlwgt'] = df['fnlwgt'].astype(int)
df['education_num'] = df['education_num'].astype(int)
df['capital_gain'] = df['capital_gain'].astype(int)
df['capital_loss'] = df['capital_loss'].astype(int)
df['hours_per_week'] = df['hours_per_week'].astype(int)

We can see that there are no missing values in the dataset.

### Types of variables


In this section, I segregate the dataset into categorical and numerical variables. There are a mixture of categorical and numerical variables in the dataset. Categorical variables have data type object. Numerical variables have data type int64.


First of all, I will explore categorical variables.

### Explore categorical variables

In [0]:
# find categorical variables

categorical = [var for var in df.columns if df[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :\n\n', categorical)

There are 9 categorical variables

The categorical variables are :

 ['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'native_country', 'income']

In [0]:
# view the categorical variables

df[categorical].head()

Out[12]:

workclass 
 education 
 marital_status 
 occupation 
 relationship 
 race 
 sex 
 native_country 
 income 
 
 
 
 
 0 
 State-gov 
 Bachelors 
 Never-married 
 Adm-clerical 
 Not-in-family 
 White 
 Male 
 United-States 
 <=50K 
 
 
 1 
 Self-emp-not-inc 
 Bachelors 
 Married-civ-spouse 
 Exec-managerial 
 Husband 
 White 
 Male 
 United-States 
 <=50K 
 
 
 2 
 Private 
 HS-grad 
 Divorced 
 Handlers-cleaners 
 Not-in-family 
 White 
 Male 
 United-States 
 <=50K 
 
 
 3 
 Private 
 11th 
 Married-civ-spouse 
 Handlers-cleaners 
 Husband 
 Black 
 Male 
 United-States 
 <=50K 
 
 
 4 
 Private 
 Bachelors 
 Married-civ-spouse 
 Prof-specialty 
 Wife 
 Black 
 Female 
 Cuba 
 <=50K

### Summary of categorical variables


- There are 9 categorical variables. 


- The categorical variables are given by `workclass`, `education`, `marital_status`, `occupation`, `relationship`, `race`, `sex`, `native_country` and `income`.


- `income` is the target variable.

### Explore problems within categorical variables


First, Lets explore the categorical variables.


### Missing values in categorical variables

In [0]:
# check missing values in categorical variables

df[categorical].isnull().sum()

Out[13]: workclass 0
education 0
marital_status 0
occupation 0
relationship 0
race 0
sex 0
native_country 0
income 0
dtype: int64

We can see that there are no missing values in the categorical variables. we shall confirm this further.

### Frequency counts of categorical variables


Now, lets check the frequency counts of categorical variables.

In [0]:
# view frequency counts of values in categorical variables

for var in categorical: 
    print(df[var].value_counts())

Private 22696
 Self-emp-not-inc 2541
 Local-gov 2093
 ? 1836
 State-gov 1298
 Self-emp-inc 1116
 Federal-gov 960
 Without-pay 14
 Never-worked 7
Name: workclass, dtype: int64
 HS-grad 10501
 Some-college 7291
 Bachelors 5355
 Masters 1723
 Assoc-voc 1382
 11th 1175
 Assoc-acdm 1067
 10th 933
 7th-8th 646
 Prof-school 576
 9th 514
 12th 433
 Doctorate 413
 5th-6th 333
 1st-4th 168
 Preschool 51
Name: education, dtype: int64
 Married-civ-spouse 14976
 Never-married 10683
 Divorced 4443
 Separated 1025
 Widowed 993
 Married-spouse-absent 418
 Married-AF-spouse 23
Name: marital_status, dtype: int64
 Prof-specialty 4140
 Craft-repair 4099
 Exec-managerial 4066
 Adm-clerical 3770
 Sales 3650
 Other-service 3295
 Machine-op-inspct 2002
 ? 1843
 Transport-moving 1597
 Handlers-cleaners 1370
 Farming-fishing 994
 Tech-support 928
 Protective-serv 649
 Priv-house-serv 149
 Armed-Forces 9
Name: occupation, dtype: int64
 Husband 13193
 Not-in-family 8305
 Own-child 5068
 Unmarried 3446
 Wife 1568
 Other-relative 981
Name: relationship, dtype: int64
 White 27816
 Black 3124
 Asian-Pac-Islander 1039
 Amer-Indian-Eskimo 311
 Other 271
Name: race, dtype: int64
 Male 21790
 Female 10771
Name: sex, dtype: int64
 United-States 29170
 Mexico 643
 ? 583
 Philippines 198
 Germany 137
 Canada 121
 Puerto-Rico 114
 El-Salvador 106
 India 100
 Cuba 95
 England 90
 Jamaica 81
 South 80
 China 75
 Italy 73
 Dominican-Republic 70
 Vietnam 67
 Guatemala 64
 Japan 62
 Poland 60
 Columbia 59
 Taiwan 51
 Haiti 44
 Iran 43
 Portugal 37
 Nicaragua 34
 Peru 31
 Greece 29
 France 29
 Ecuador 28
 Ireland 24
 Hong 20
 Cambodia 19
 Trinadad&Tobago 19
 Thailand 18
 Laos 18
 Yugoslavia 16
 Outlying-US(Guam-USVI-etc) 14
 Honduras 13
 Hungary 13
 Scotland 12
 Holand-Netherlands 1
Name: native_country, dtype: int64
 <=50K 24720
 >50K 7841
Name: income, dtype: int64

In [0]:
# view frequency distribution of categorical variables

for var in categorical: 
    print(df[var].value_counts()/np.float(len(df)))

Private 0.697030
 Self-emp-not-inc 0.078038
 Local-gov 0.064279
 ? 0.056386
 State-gov 0.039864
 Self-emp-inc 0.034274
 Federal-gov 0.029483
 Without-pay 0.000430
 Never-worked 0.000215
Name: workclass, dtype: float64
 HS-grad 0.322502
 Some-college 0.223918
 Bachelors 0.164461
 Masters 0.052916
 Assoc-voc 0.042443
 11th 0.036086
 Assoc-acdm 0.032769
 10th 0.028654
 7th-8th 0.019840
 Prof-school 0.017690
 9th 0.015786
 12th 0.013298
 Doctorate 0.012684
 5th-6th 0.010227
 1st-4th 0.005160
 Preschool 0.001566
Name: education, dtype: float64
 Married-civ-spouse 0.459937
 Never-married 0.328092
 Divorced 0.136452
 Separated 0.031479
 Widowed 0.030497
 Married-spouse-absent 0.012837
 Married-AF-spouse 0.000706
Name: marital_status, dtype: float64
 Prof-specialty 0.127146
 Craft-repair 0.125887
 Exec-managerial 0.124873
 Adm-clerical 0.115783
 Sales 0.112097
 Other-service 0.101195
 Machine-op-inspct 0.061485
 ? 0.056601
 Transport-moving 0.049046
 Handlers-cleaners 0.042075
 Farming-fishing 0.030527
 Tech-support 0.028500
 Protective-serv 0.019932
 Priv-house-serv 0.004576
 Armed-Forces 0.000276
Name: occupation, dtype: float64
 Husband 0.405178
 Not-in-family 0.255060
 Own-child 0.155646
 Unmarried 0.105832
 Wife 0.048156
 Other-relative 0.030128
Name: relationship, dtype: float64
 White 0.854274
 Black 0.095943
 Asian-Pac-Islander 0.031909
 Amer-Indian-Eskimo 0.009551
 Other 0.008323
Name: race, dtype: float64
 Male 0.669205
 Female 0.330795
Name: sex, dtype: float64
 United-States 0.895857
 Mexico 0.019748
 ? 0.017905
 Philippines 0.006081
 Germany 0.004207
 Canada 0.003716
 Puerto-Rico 0.003501
 El-Salvador 0.003255
 India 0.003071
 Cuba 0.002918
 England 0.002764
 Jamaica 0.002488
 South 0.002457
 China 0.002303
 Italy 0.002242
 Dominican-Republic 0.002150
 Vietnam 0.002058
 Guatemala 0.001966
 Japan 0.001904
 Poland 0.001843
 Columbia 0.001812
 Taiwan 0.001566
 Haiti 0.001351
 Iran 0.001321
 Portugal 0.001136
 Nicaragua 0.001044
 Peru 0.000952
 Greece 0.000891
 France 0.000891
 Ecuador 0.000860
 Ireland 0.000737
 Hong 0.000614
 Cambodia 0.000584
 Trinadad&Tobago 0.000584
 Thailand 0.000553
 Laos 0.000553
 Yugoslavia 0.000491
 Outlying-US(Guam-USVI-etc) 0.000430
 Honduras 0.000399
 Hungary 0.000399
 Scotland 0.000369
 Holand-Netherlands 0.000031
Name: native_country, dtype: float64
 <=50K 0.75919
 >50K 0.24081
Name: income, dtype: float64

Now, we can see that there are several variables like `workclass`, `occupation` and `native_country` which contain missing values. Generally, the missing values are coded as `NaN` and python will detect them with the usual command of `df.isnull().sum()`.

But, in this case the missing values are coded as `?`. Python fail to detect these as missing values because it do not consider `?` as missing values. So, I have to replace `?` with `NaN` so that Python can detect these missing values.

Lets explore these variables and replace `?` with `NaN`.

### Explore workclass variable

In [0]:
# check labels in workclass variable

df.workclass.unique()

Out[16]: array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
 ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay',
 ' Never-worked'], dtype=object)

In [0]:
# check frequency distribution of values in workclass variable

df.workclass.value_counts()

Out[17]: Private 22696
 Self-emp-not-inc 2541
 Local-gov 2093
 ? 1836
 State-gov 1298
 Self-emp-inc 1116
 Federal-gov 960
 Without-pay 14
 Never-worked 7
Name: workclass, dtype: int64

We can see that there are 1836 values encoded as `?` in workclass variable. we shall replace these `?` with `NaN`.

In [0]:
# replace '?' values in workclass variable with `NaN`


df['workclass'].replace('?', np.NaN, inplace=True)

In [0]:
# again check the frequency distribution of values in workclass variable

df.workclass.value_counts()

Out[19]: Private 22696
 Self-emp-not-inc 2541
 Local-gov 2093
 ? 1836
 State-gov 1298
 Self-emp-inc 1116
 Federal-gov 960
 Without-pay 14
 Never-worked 7
Name: workclass, dtype: int64

Now, we can see that there are no values encoded as `?` in the `workclass` variable.

I will adopt similar approach with `occupation` and `native_country` column.

### Explore occupation variable

In [0]:
# check labels in occupation variable

df.occupation.unique()

Out[20]: array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
 ' Prof-specialty', ' Other-service', ' Sales', ' Craft-repair',
 ' Transport-moving', ' Farming-fishing', ' Machine-op-inspct',
 ' Tech-support', ' ?', ' Protective-serv', ' Armed-Forces',
 ' Priv-house-serv'], dtype=object)

In [0]:
# check frequency distribution of values in occupation variable

df.occupation.value_counts()

Out[21]: Prof-specialty 4140
 Craft-repair 4099
 Exec-managerial 4066
 Adm-clerical 3770
 Sales 3650
 Other-service 3295
 Machine-op-inspct 2002
 ? 1843
 Transport-moving 1597
 Handlers-cleaners 1370
 Farming-fishing 994
 Tech-support 928
 Protective-serv 649
 Priv-house-serv 149
 Armed-Forces 9
Name: occupation, dtype: int64

We can see that there are 1843 values encoded as `?` in `occupation` variable. we shall replace these `?` with `NaN`.

In [0]:
# replace '?' values in occupation variable with `NaN`

df['occupation'].replace('?', np.NaN, inplace=True)

In [0]:
# again check the frequency distribution of values in occupation variable

df.occupation.value_counts()

Out[23]: Prof-specialty 4140
 Craft-repair 4099
 Exec-managerial 4066
 Adm-clerical 3770
 Sales 3650
 Other-service 3295
 Machine-op-inspct 2002
 ? 1843
 Transport-moving 1597
 Handlers-cleaners 1370
 Farming-fishing 994
 Tech-support 928
 Protective-serv 649
 Priv-house-serv 149
 Armed-Forces 9
Name: occupation, dtype: int64

### Explore native_country variable

In [0]:
# check labels in native_country variable
df.native_country.unique()

Out[24]: array([' United-States', ' Cuba', ' Jamaica', ' India', ' ?', ' Mexico',
 ' South', ' Puerto-Rico', ' Honduras', ' England', ' Canada',
 ' Germany', ' Iran', ' Philippines', ' Italy', ' Poland',
 ' Columbia', ' Cambodia', ' Thailand', ' Ecuador', ' Laos',
 ' Taiwan', ' Haiti', ' Portugal', ' Dominican-Republic',
 ' El-Salvador', ' France', ' Guatemala', ' China', ' Japan',
 ' Yugoslavia', ' Peru', ' Outlying-US(Guam-USVI-etc)', ' Scotland',
 ' Trinadad&Tobago', ' Greece', ' Nicaragua', ' Vietnam', ' Hong',
 ' Ireland', ' Hungary', ' Holand-Netherlands'], dtype=object)

In [0]:
# check frequency distribution of values in native_country variable
df.native_country.value_counts()

Out[25]: United-States 29170
 Mexico 643
 ? 583
 Philippines 198
 Germany 137
 Canada 121
 Puerto-Rico 114
 El-Salvador 106
 India 100
 Cuba 95
 England 90
 Jamaica 81
 South 80
 China 75
 Italy 73
 Dominican-Republic 70
 Vietnam 67
 Guatemala 64
 Japan 62
 Poland 60
 Columbia 59
 Taiwan 51
 Haiti 44
 Iran 43
 Portugal 37
 Nicaragua 34
 Peru 31
 Greece 29
 France 29
 Ecuador 28
 Ireland 24
 Hong 20
 Cambodia 19
 Trinadad&Tobago 19
 Thailand 18
 Laos 18
 Yugoslavia 16
 Outlying-US(Guam-USVI-etc) 14
 Honduras 13
 Hungary 13
 Scotland 12
 Holand-Netherlands 1
Name: native_country, dtype: int64

We can see that there are 583 values encoded as `?` in `native_country` variable. I will replace these `?` with `NaN`.

In [0]:
# replace '?' values in native_country variable with `NaN`

df['native_country'].replace('?', np.NaN, inplace=True)

In [0]:
# again check the frequency distribution of values in native_country variable

df.native_country.value_counts()

Out[27]: United-States 29170
 Mexico 643
 ? 583
 Philippines 198
 Germany 137
 Canada 121
 Puerto-Rico 114
 El-Salvador 106
 India 100
 Cuba 95
 England 90
 Jamaica 81
 South 80
 China 75
 Italy 73
 Dominican-Republic 70
 Vietnam 67
 Guatemala 64
 Japan 62
 Poland 60
 Columbia 59
 Taiwan 51
 Haiti 44
 Iran 43
 Portugal 37
 Nicaragua 34
 Peru 31
 Greece 29
 France 29
 Ecuador 28
 Ireland 24
 Hong 20
 Cambodia 19
 Trinadad&Tobago 19
 Thailand 18
 Laos 18
 Yugoslavia 16
 Outlying-US(Guam-USVI-etc) 14
 Honduras 13
 Hungary 13
 Scotland 12
 Holand-Netherlands 1
Name: native_country, dtype: int64

### Check missing values in categorical variables again

In [0]:
df[categorical].isnull().sum()

Out[28]: workclass 0
education 0
marital_status 0
occupation 0
relationship 0
race 0
sex 0
native_country 0
income 0
dtype: int64

Now, we can see that `workclass`, `occupation` and `native_country` variable contains missing values.

### Number of labels: cardinality


The number of labels within a categorical variable is known as **cardinality**. A high number of labels within a variable is known as **high cardinality**. High cardinality may pose some serious problems in the machine learning model. So, Lets check for high cardinality.

In [0]:
# check for cardinality in categorical variables

for var in categorical:
    print(var, ' contains ', len(df[var].unique()), ' labels')

workclass contains 9 labels
education contains 16 labels
marital_status contains 7 labels
occupation contains 15 labels
relationship contains 6 labels
race contains 5 labels
sex contains 2 labels
native_country contains 42 labels
income contains 2 labels

We can see that `native_country` column contains relatively large number of labels as compared to other columns. we shall check for cardinality after train-test split.

### Explore Numerical Variables

In [0]:
# find numerical variables

numerical = [var for var in df.columns if df[var].dtype!='O']

print('There are {} numerical variables\n'.format(len(numerical)))

print('The numerical variables are :', numerical)

There are 6 numerical variables

The numerical variables are : ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

In [0]:
# view the numerical variables

df[numerical].head()

Out[31]:

age 
 fnlwgt 
 education_num 
 capital_gain 
 capital_loss 
 hours_per_week 
 
 
 
 
 0 
 39 
 77516 
 13 
 2174 
 0 
 40 
 
 
 1 
 50 
 83311 
 13 
 0 
 0 
 13 
 
 
 2 
 38 
 215646 
 9 
 0 
 0 
 40 
 
 
 3 
 53 
 234721 
 7 
 0 
 0 
 40 
 
 
 4 
 28 
 338409 
 13 
 0 
 0 
 40

### Summary of numerical variables


- There are 6 numerical variables. 


- These are given by `age`, `fnlwgt`, `education_num`, `capital_gain`, `capital_loss` and `hours_per_week`.


- All of the numerical variables are of discrete data type.

### Explore problems within numerical variables


Now, Lets  explore the numerical variables.


### Missing values in numerical variables

In [0]:
# check missing values in numerical variables

df[numerical].isnull().sum()

Out[32]: age 0
fnlwgt 0
education_num 0
capital_gain 0
capital_loss 0
hours_per_week 0
dtype: int64

We can see that all the 6 numerical variables do not contain missing values.

# **Declare feature vector and target variable** <a class="anchor" id="8"></a>

In [0]:
X = df.drop(['income'], axis=1)

y = df['income']

# **Split data into separate training and test set** <a class="anchor" id="9"></a>

[Table of Contents](#0.1)

In [0]:
# split X and y into training and testing sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)

In [0]:

# check the shape of X_train and X_test

X_train.shape, X_test.shape

Out[35]: ((22792, 14), (9769, 14))

# **Feature Engineering** <a class="anchor" id="10"></a>

**Feature Engineering** is the process of transforming raw data into useful features that help us to understand our model better and increase its predictive power. Lets carry out feature engineering on different types of variables.


First, We shall see the categorical and numerical variables again separately.

In [0]:
# check data types in X_train

X_train.dtypes

Out[36]: age int64
workclass object
fnlwgt int64
education object
education_num int64
marital_status object
occupation object
relationship object
race object
sex object
capital_gain int64
capital_loss int64
hours_per_week int64
native_country object
dtype: object

In [0]:
# display categorical variables

categorical = [col for col in X_train.columns if X_train[col].dtypes == 'O']

categorical

Out[37]: ['workclass',
 'education',
 'marital_status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native_country']

In [0]:
# display numerical variables

numerical = [col for col in X_train.columns if X_train[col].dtypes != 'O']

numerical

Out[38]: ['age',
 'fnlwgt',
 'education_num',
 'capital_gain',
 'capital_loss',
 'hours_per_week']

### Engineering missing values in categorical variables

In [0]:
# print percentage of missing values in the categorical variables in training set

X_train[categorical].isnull().mean()

Out[39]: workclass 0.0
education 0.0
marital_status 0.0
occupation 0.0
relationship 0.0
race 0.0
sex 0.0
native_country 0.0
dtype: float64

In [0]:
# print categorical variables with missing data

for col in categorical:
    if X_train[col].isnull().mean()>0:
        print(col, (X_train[col].isnull().mean()))

In [0]:
# impute missing categorical variables with most frequent value

for df2 in [X_train, X_test]:
    df2['workclass'].fillna(X_train['workclass'].mode()[0], inplace=True)
    df2['occupation'].fillna(X_train['occupation'].mode()[0], inplace=True)
    df2['native_country'].fillna(X_train['native_country'].mode()[0], inplace=True)    

In [0]:
# check missing values in categorical variables in X_train
X_train[categorical].isnull().sum()

Out[42]: workclass 0
education 0
marital_status 0
occupation 0
relationship 0
race 0
sex 0
native_country 0
dtype: int64

In [0]:
# check missing values in categorical variables in X_test

X_test[categorical].isnull().sum()

Out[43]: workclass 0
education 0
marital_status 0
occupation 0
relationship 0
race 0
sex 0
native_country 0
dtype: int64

As a final check, Lets check for missing values in X_train and X_test.

In [0]:
# check missing values in X_train

X_train.isnull().sum()

Out[44]: age 0
workclass 0
fnlwgt 0
education 0
education_num 0
marital_status 0
occupation 0
relationship 0
race 0
sex 0
capital_gain 0
capital_loss 0
hours_per_week 0
native_country 0
dtype: int64

In [0]:
# check missing values in X_test

X_test.isnull().sum()

Out[45]: age 0
workclass 0
fnlwgt 0
education 0
education_num 0
marital_status 0
occupation 0
relationship 0
race 0
sex 0
capital_gain 0
capital_loss 0
hours_per_week 0
native_country 0
dtype: int64

We can see that there are no missing values in X_train and X_test.

### Encode categorical variables

In [0]:
# print categorical variables

categorical

Out[46]: ['workclass',
 'education',
 'marital_status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native_country']

In [0]:
X_train[categorical].head()

Out[47]:

workclass 
 education 
 marital_status 
 occupation 
 relationship 
 race 
 sex 
 native_country 
 
 
 
 
 32098 
 Private 
 HS-grad 
 Married-civ-spouse 
 Craft-repair 
 Husband 
 White 
 Male 
 United-States 
 
 
 25206 
 State-gov 
 HS-grad 
 Divorced 
 Adm-clerical 
 Unmarried 
 White 
 Female 
 United-States 
 
 
 23491 
 Private 
 Some-college 
 Married-civ-spouse 
 Sales 
 Husband 
 White 
 Male 
 United-States 
 
 
 12367 
 Private 
 HS-grad 
 Never-married 
 Craft-repair 
 Not-in-family 
 White 
 Male 
 Guatemala 
 
 
 7054 
 Private 
 7th-8th 
 Never-married 
 Craft-repair 
 Not-in-family 
 White 
 Male 
 Germany

In [0]:
X_train['workclass'].unique()

Out[48]: array([' Private', ' State-gov', ' Self-emp-not-inc', ' ?',
 ' Federal-gov', ' Self-emp-inc', ' Local-gov', ' Without-pay',
 ' Never-worked'], dtype=object)

In [0]:
len(X_train['native_country'].unique())

Out[49]: 42

In [0]:
!pip install category_encoders

Requirement already satisfied: category_encoders in /databricks/python3/lib/python3.8/site-packages (2.2.2)
Requirement already satisfied: numpy>=1.14.0 in /databricks/python3/lib/python3.8/site-packages (from category_encoders) (1.19.2)
Requirement already satisfied: patsy>=0.5.1 in /databricks/python3/lib/python3.8/site-packages (from category_encoders) (0.5.1)
Requirement already satisfied: scikit-learn>=0.20.0 in /databricks/python3/lib/python3.8/site-packages (from category_encoders) (0.23.2)
Requirement already satisfied: statsmodels>=0.9.0 in /databricks/python3/lib/python3.8/site-packages (from category_encoders) (0.12.0)
Requirement already satisfied: pandas>=0.21.1 in /databricks/python3/lib/python3.8/site-packages (from category_encoders) (1.1.5)
Requirement already satisfied: scipy>=1.0.0 in /databricks/python3/lib/python3.8/site-packages (from category_encoders) (1.5.2)
Requirement already satisfied: six in /databricks/python3/lib/python3.8/site-packages (from patsy>=0.5.1->category_encoders) (1.15.0)
Requirement already satisfied: threadpoolctl>=2.0.0 in /databricks/python3/lib/python3.8/site-packages (from scikit-learn>=0.20.0->category_encoders) (2.1.0)
Requirement already satisfied: joblib>=0.11 in /databricks/python3/lib/python3.8/site-packages (from scikit-learn>=0.20.0->category_encoders) (0.17.0)
Requirement already satisfied: pytz>=2017.2 in /databricks/python3/lib/python3.8/site-packages (from pandas>=0.21.1->category_encoders) (2020.5)
Requirement already satisfied: python-dateutil>=2.7.3 in /databricks/python3/lib/python3.8/site-packages (from pandas>=0.21.1->category_encoders) (2.8.1)
 WARNING: You are using pip version 20.2.4; however, version 21.2.4 is available.
You should consider upgrading via the '/databricks/python3/bin/python -m pip install --upgrade pip' command.

In [0]:
# import category encoders

import category_encoders as ce

In [0]:
# encode remaining variables with one-hot encoding

encoder = ce.OneHotEncoder(cols=['workclass', 'education', 'marital_status', 'occupation', 'relationship', 
                                 'race', 'sex', 'native_country'])

X_train = encoder.fit_transform(X_train)

X_test = encoder.transform(X_test)

In [0]:
X_train.head()

Out[55]:

age 
 workclass_1 
 workclass_2 
 workclass_3 
 workclass_4 
 workclass_5 
 workclass_6 
 workclass_7 
 workclass_8 
 workclass_9 
 fnlwgt 
 education_1 
 education_2 
 education_3 
 education_4 
 education_5 
 education_6 
 education_7 
 education_8 
 education_9 
 education_10 
 education_11 
 education_12 
 education_13 
 education_14 
 education_15 
 education_16 
 education_num 
 marital_status_1 
 marital_status_2 
 marital_status_3 
 marital_status_4 
 marital_status_5 
 marital_status_6 
 marital_status_7 
 occupation_1 
 occupation_2 
 occupation_3 
 occupation_4 
 occupation_5 
 ... 
 native_country_3 
 native_country_4 
 native_country_5 
 native_country_6 
 native_country_7 
 native_country_8 
 native_country_9 
 native_country_10 
 native_country_11 
 native_country_12 
 native_country_13 
 native_country_14 
 native_country_15 
 native_country_16 
 native_country_17 
 native_country_18 
 native_country_19 
 native_country_20 
 native_country_21 
 native_country_22 
 native_country_23 
 native_country_24 
 native_country_25 
 native_country_26 
 native_country_27 
 native_country_28 
 native_country_29 
 native_country_30 
 native_country_31 
 native_country_32 
 native_country_33 
 native_country_34 
 native_country_35 
 native_country_36 
 native_country_37 
 native_country_38 
 native_country_39 
 native_country_40 
 native_country_41 
 native_country_42 
 
 
 
 
 32098 
 45 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 170871 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 9 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 25206 
 47 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 108890 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 9 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 23491 
 48 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 187505 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 10 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 12367 
 29 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 145592 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 9 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 7054 
 23 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 203003 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 4 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 ... 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 
 5 rows × 108 columns

In [0]:
X_train.shape

Out[56]: (22792, 108)

We can see that from the initial 14 columns, we now have 113 columns.

Similarly,We will take a look at the `X_test` set.

In [0]:
X_test.head()

Out[57]:

age 
 workclass_1 
 workclass_2 
 workclass_3 
 workclass_4 
 workclass_5 
 workclass_6 
 workclass_7 
 workclass_8 
 workclass_9 
 fnlwgt 
 education_1 
 education_2 
 education_3 
 education_4 
 education_5 
 education_6 
 education_7 
 education_8 
 education_9 
 education_10 
 education_11 
 education_12 
 education_13 
 education_14 
 education_15 
 education_16 
 education_num 
 marital_status_1 
 marital_status_2 
 marital_status_3 
 marital_status_4 
 marital_status_5 
 marital_status_6 
 marital_status_7 
 occupation_1 
 occupation_2 
 occupation_3 
 occupation_4 
 occupation_5 
 ... 
 native_country_3 
 native_country_4 
 native_country_5 
 native_country_6 
 native_country_7 
 native_country_8 
 native_country_9 
 native_country_10 
 native_country_11 
 native_country_12 
 native_country_13 
 native_country_14 
 native_country_15 
 native_country_16 
 native_country_17 
 native_country_18 
 native_country_19 
 native_country_20 
 native_country_21 
 native_country_22 
 native_country_23 
 native_country_24 
 native_country_25 
 native_country_26 
 native_country_27 
 native_country_28 
 native_country_29 
 native_country_30 
 native_country_31 
 native_country_32 
 native_country_33 
 native_country_34 
 native_country_35 
 native_country_36 
 native_country_37 
 native_country_38 
 native_country_39 
 native_country_40 
 native_country_41 
 native_country_42 
 
 
 
 
 22278 
 27 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 177119 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 10 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 8950 
 27 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 216481 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 13 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 7838 
 25 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 256263 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 12 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 16505 
 46 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 147640 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 3 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 19140 
 45 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 172822 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 7 
 0 
 1 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 ... 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 0 
 
 
 
 5 rows × 108 columns

In [0]:
X_test.shape

Out[58]: (9769, 108)

We now have training and testing set ready for model building. Before that, we should map all the feature variables onto the same scale. It is called `feature scaling`. I will do it as follows.

# **Feature Scaling** <a class="anchor" id="11"></a>

In [0]:
cols = X_train.columns

In [0]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X_train = scaler.fit_transform(X_train)

X_test = scaler.transform(X_test)


In [0]:
X_train = pd.DataFrame(X_train, columns=[cols])

In [0]:
X_test = pd.DataFrame(X_test, columns=[cols])

In [0]:
X_train.head()

Out[63]:

age 
 workclass_1 
 workclass_2 
 workclass_3 
 workclass_4 
 workclass_5 
 workclass_6 
 workclass_7 
 workclass_8 
 workclass_9 
 fnlwgt 
 education_1 
 education_2 
 education_3 
 education_4 
 education_5 
 education_6 
 education_7 
 education_8 
 education_9 
 education_10 
 education_11 
 education_12 
 education_13 
 education_14 
 education_15 
 education_16 
 education_num 
 marital_status_1 
 marital_status_2 
 marital_status_3 
 marital_status_4 
 marital_status_5 
 marital_status_6 
 marital_status_7 
 occupation_1 
 occupation_2 
 occupation_3 
 occupation_4 
 occupation_5 
 ... 
 native_country_3 
 native_country_4 
 native_country_5 
 native_country_6 
 native_country_7 
 native_country_8 
 native_country_9 
 native_country_10 
 native_country_11 
 native_country_12 
 native_country_13 
 native_country_14 
 native_country_15 
 native_country_16 
 native_country_17 
 native_country_18 
 native_country_19 
 native_country_20 
 native_country_21 
 native_country_22 
 native_country_23 
 native_country_24 
 native_country_25 
 native_country_26 
 native_country_27 
 native_country_28 
 native_country_29 
 native_country_30 
 native_country_31 
 native_country_32 
 native_country_33 
 native_country_34 
 native_country_35 
 native_country_36 
 native_country_37 
 native_country_38 
 native_country_39 
 native_country_40 
 native_country_41 
 native_country_42 
 
 
 
 
 0 
 0.40 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.058906 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.333333 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 1 
 0.50 
 -1.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.578076 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.333333 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 2 
 0.55 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.080425 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.000000 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 3 
 -0.40 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.270650 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.333333 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 4 
 -0.70 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.210240 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -2.000000 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 
 5 rows × 108 columns

We now have `X_train` dataset ready to be fed into the Gaussian Naive Bayes classifier. I will do it as follows.

In [0]:
y_train

Out[64]: 32098 >50K
25206 <=50K
23491 >50K
12367 <=50K
7054 <=50K
 ... 
13123 >50K
19648 <=50K
9845 <=50K
10799 >50K
2732 <=50K
Name: income, Length: 22792, dtype: object

In [0]:
y_test

Out[65]: 22278 <=50K
8950 <=50K
7838 <=50K
16505 <=50K
19140 >50K
 ... 
21949 >50K
26405 >50K
23236 >50K
26823 <=50K
20721 <=50K
Name: income, Length: 9769, dtype: object

In [0]:
print(X_train.shape)
print(X_test.shape)

(22792, 108)
(9769, 108)

In [0]:
X_test.head()

Out[67]:

age 
 workclass_1 
 workclass_2 
 workclass_3 
 workclass_4 
 workclass_5 
 workclass_6 
 workclass_7 
 workclass_8 
 workclass_9 
 fnlwgt 
 education_1 
 education_2 
 education_3 
 education_4 
 education_5 
 education_6 
 education_7 
 education_8 
 education_9 
 education_10 
 education_11 
 education_12 
 education_13 
 education_14 
 education_15 
 education_16 
 education_num 
 marital_status_1 
 marital_status_2 
 marital_status_3 
 marital_status_4 
 marital_status_5 
 marital_status_6 
 marital_status_7 
 occupation_1 
 occupation_2 
 occupation_3 
 occupation_4 
 occupation_5 
 ... 
 native_country_3 
 native_country_4 
 native_country_5 
 native_country_6 
 native_country_7 
 native_country_8 
 native_country_9 
 native_country_10 
 native_country_11 
 native_country_12 
 native_country_13 
 native_country_14 
 native_country_15 
 native_country_16 
 native_country_17 
 native_country_18 
 native_country_19 
 native_country_20 
 native_country_21 
 native_country_22 
 native_country_23 
 native_country_24 
 native_country_25 
 native_country_26 
 native_country_27 
 native_country_28 
 native_country_29 
 native_country_30 
 native_country_31 
 native_country_32 
 native_country_33 
 native_country_34 
 native_country_35 
 native_country_36 
 native_country_37 
 native_country_38 
 native_country_39 
 native_country_40 
 native_country_41 
 native_country_42 
 
 
 
 
 0 
 -0.50 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.006571 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.000000 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 1 
 -0.50 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.323136 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.000000 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 2 
 -0.60 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.656361 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.666667 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 3 
 0.45 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.253496 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 -2.333333 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 4 
 0.40 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.042564 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -1.000000 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 
 5 rows × 108 columns

In [0]:
X_train.head()

Out[68]:

age 
 workclass_1 
 workclass_2 
 workclass_3 
 workclass_4 
 workclass_5 
 workclass_6 
 workclass_7 
 workclass_8 
 workclass_9 
 fnlwgt 
 education_1 
 education_2 
 education_3 
 education_4 
 education_5 
 education_6 
 education_7 
 education_8 
 education_9 
 education_10 
 education_11 
 education_12 
 education_13 
 education_14 
 education_15 
 education_16 
 education_num 
 marital_status_1 
 marital_status_2 
 marital_status_3 
 marital_status_4 
 marital_status_5 
 marital_status_6 
 marital_status_7 
 occupation_1 
 occupation_2 
 occupation_3 
 occupation_4 
 occupation_5 
 ... 
 native_country_3 
 native_country_4 
 native_country_5 
 native_country_6 
 native_country_7 
 native_country_8 
 native_country_9 
 native_country_10 
 native_country_11 
 native_country_12 
 native_country_13 
 native_country_14 
 native_country_15 
 native_country_16 
 native_country_17 
 native_country_18 
 native_country_19 
 native_country_20 
 native_country_21 
 native_country_22 
 native_country_23 
 native_country_24 
 native_country_25 
 native_country_26 
 native_country_27 
 native_country_28 
 native_country_29 
 native_country_30 
 native_country_31 
 native_country_32 
 native_country_33 
 native_country_34 
 native_country_35 
 native_country_36 
 native_country_37 
 native_country_38 
 native_country_39 
 native_country_40 
 native_country_41 
 native_country_42 
 
 
 
 
 0 
 0.40 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.058906 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.333333 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 1 
 0.50 
 -1.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.578076 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.333333 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 2 
 0.55 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.080425 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.000000 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 3 
 -0.40 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.270650 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -0.333333 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 4 
 -0.70 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.210240 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 -2.000000 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 ... 
 1.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 0.0 
 
 
 
 5 rows × 108 columns

In [0]:
X_test.columns = [col[0] for col in X_test.columns]

In [0]:
X_test.columns

Out[70]: Index(['age', 'workclass_1', 'workclass_2', 'workclass_3', 'workclass_4',
 'workclass_5', 'workclass_6', 'workclass_7', 'workclass_8',
 'workclass_9',
 ...
 'native_country_33', 'native_country_34', 'native_country_35',
 'native_country_36', 'native_country_37', 'native_country_38',
 'native_country_39', 'native_country_40', 'native_country_41',
 'native_country_42'],
 dtype='object', length=108)

In [0]:
sub_df = X_test.iloc[500:1000]
s_df = spark.createDataFrame(sub_df)

In [0]:
display(s_df)

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5 workclass_6 workclass_7 workclass_8 workclass_9 fnlwgt education_1 education_2 education_3 education_4 education_5 education_6 education_7 education_8 education_9 education_10 education_11 education_12 education_13 education_14 education_15 education_16 education_num marital_status_1 marital_status_2 marital_status_3 marital_status_4 marital_status_5 marital_status_6 marital_status_7 occupation_1 occupation_2 occupation_3 occupation_4 occupation_5 occupation_6 occupation_7 occupation_8 occupation_9 occupation_10 occupation_11 occupation_12 occupation_13 occupation_14 occupation_15 relationship_1 relationship_2 relationship_3 relationship_4 relationship_5 relationship_6 race_1 race_2 race_3 race_4 race_5 sex_1 sex_2 capital_gain capital_loss hours_per_week native_country_1 native_country_2 native_country_3 native_country_4 native_country_5 native_country_6 native_country_7 native_country_8 native_country_9 native_country_10 native_country_11 native_country_12 native_country_13 native_country_14 native_country_15 native_country_16 native_country_17 native_country_18 native_country_19 native_country_20 native_country_21 native_country_22 native_country_23 native_country_24 native_country_25 native_country_26 native_country_27 native_country_28 native_country_29 native_country_30 native_country_31 native_country_32 native_country_33 native_country_34 native_country_35 native_country_36 native_country_37 native_country_38 native_country_39 native_country_40 native_country_41 native_country_42 0.05 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.524049344660855 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.3333333333333333 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 -1.0 0.0 1.0 0.0 0.0 -1.0 1.0 0.0 0.0 -1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.25 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.4895076632484467 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.3333333333333333 0.0 0.0 1.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.8 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -0.65 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -0.763569048810673 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 -1.0 1.0 0.0 0.0 0.0 -1.0 1.0 0.0 0.0 -4.8 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.7 -1.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 -0.23887054251066406 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -0.3333333333333333 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -6.2 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.85 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -0.7061077733965184 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -0.3333333333333333 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1848.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 3.2219734932644246 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0

In [0]:
s_df.write.mode('overwrite').saveAsTable('nb_testdata')

# Check the column names of train

In [0]:
X_train.columns = [col[0] for col in X_train.columns]

In [0]:
X_train.columns

Out[75]: Index(['age', 'workclass_1', 'workclass_2', 'workclass_3', 'workclass_4',
 'workclass_5', 'workclass_6', 'workclass_7', 'workclass_8',
 'workclass_9',
 ...
 'native_country_33', 'native_country_34', 'native_country_35',
 'native_country_36', 'native_country_37', 'native_country_38',
 'native_country_39', 'native_country_40', 'native_country_41',
 'native_country_42'],
 dtype='object', length=108)

In [0]:
y_test

Out[76]: 22278 <=50K
8950 <=50K
7838 <=50K
16505 <=50K
19140 >50K
 ... 
21949 >50K
26405 >50K
23236 >50K
26823 <=50K
20721 <=50K
Name: income, Length: 9769, dtype: object

In [0]:
y_train

Out[77]: 32098 >50K
25206 <=50K
23491 >50K
12367 <=50K
7054 <=50K
 ... 
13123 >50K
19648 <=50K
9845 <=50K
10799 >50K
2732 <=50K
Name: income, Length: 22792, dtype: object

# **Model training and Logging the Experiments**

In [0]:
#Necessary imports for mlflow
import mlflow
import mlflow.pyfunc
import mlflow.sklearn
from mlflow.models.signature import infer_signature
from mlflow.utils.environment import _mlflow_conda_env
import os
# train a Gaussian Naive Bayes classifier on the training set
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_auc_score

## Naive Bayes Experiment

In [0]:
with mlflow.start_run(run_name = 'Naive Bayes'):
  # instantiate the model
  gnb = GaussianNB()
  # fit the model
  gnb.fit(X_train, y_train)
  
  # Make predictions
  predictions = gnb.predict(X_test) #Prediction on test dataset
  y_pred_train = gnb.predict(X_train) #Prediction on train dataset
  
  #Train accuracy
  train_accuracy = accuracy_score(y_train, y_pred_train)
  
  #Test accuracy
  test_accuracy = accuracy_score(y_test, predictions)
  #AUC Calculation
#   ROC_AUC = roc_auc_score(y_test, predictions)
   
  #Test Precision and Recall calculation
  cm = confusion_matrix(y_test, predictions)
  TP = cm[0,0]
  TN = cm[1,1]
  FP = cm[0,1]
  FN = cm[1,0]
  
  #Precision score
  precision = TP / float(TP + FP)
  
  #Recall
  recall = TP / float(TP + FN)
  
  # Log parameters to be tracked for comparisons
  mlflow.log_param("priors", gnb.get_params()['priors'])
  mlflow.log_param("var_smoothing", gnb.get_params()['var_smoothing'])
  
  #Infer schema and make it as signature
  signature = infer_signature(X_train, gnb.predict(X_train))

  # Log model
  mlflow.sklearn.log_model(gnb, "naive-bayes-model",  signature=signature)
  
    
  # Log metrics
  mlflow.log_metric("Training Accuracy", train_accuracy)
  mlflow.log_metric("Test Accuracy", test_accuracy)
  mlflow.log_metric("Precision", precision)
  mlflow.log_metric("Recall", recall)

## Random Forest Classifier

In [0]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

with mlflow.start_run(run_name = 'Random Forest Classifier'):
  # instantiate the model
  clf=RandomForestClassifier(n_estimators=100)
  # fit the model
  param_dist = {'max_depth': [2, 3, 4],
              'bootstrap': [True, False],
              'max_features': ['auto', 'sqrt', 'log2', None],
              'criterion': ['gini', 'entropy']
               }
  cv_rf = GridSearchCV(clf, cv = 10, param_grid=param_dist, n_jobs = 3)
  cv_rf.fit(X_train, y_train)
  
  # Make predictions
  predictions = cv_rf.predict(X_test) #Prediction on test dataset
  y_pred_train = cv_rf.predict(X_train) #Prediction on train dataset
  
  #Train accuracy
  train_accuracy = accuracy_score(y_train, y_pred_train)
  
  #Test accuracy
  test_accuracy = accuracy_score(y_test, predictions)
  #AUC Calculation
#   ROC_AUC = roc_auc_score(y_test, predictions)
   
  #Test Precision and Recall calculation
  cm = confusion_matrix(y_test, predictions)
  TP = cm[0,0]
  TN = cm[1,1]
  FP = cm[0,1]
  FN = cm[1,0]
  
  #Precision score
  precision = TP / float(TP + FP)
  
  #Recall
  recall = TP / float(TP + FN)
  
  # Log parameters to be tracked for comparisons
  mlflow.log_param("boostrap", cv_rf.best_params_['bootstrap'])
  mlflow.log_param("max_depth", cv_rf.best_params_['max_depth'])
  mlflow.log_param("max_features", cv_rf.best_params_['max_features'])
  mlflow.log_param("criterion", cv_rf.best_params_['criterion'])
  
  #Infer schema and make it as signature
  signature = infer_signature(X_train, cv_rf.predict(X_train))

  # Log model with signature of the schema of the model
  mlflow.sklearn.log_model(cv_rf, "random-forest-model",  signature=signature)
  
    
  # Log metrics
  mlflow.log_metric("Training Accuracy", train_accuracy)
  mlflow.log_metric("Test Accuracy", test_accuracy)
  mlflow.log_metric("Precision", precision)
  mlflow.log_metric("Recall", recall)